src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -81,173 +81,229 @@ CHIP_ERROR SimpleSubscriptionResumptionStorage::SaveIndex(const SubscriptionInde
+            return __err;                                                                                                          \
+        }                                                                                                                          \
+    } while (false)
+
 CHIP_ERROR SimpleSubscriptionResumptionStorage::LoadIndex(SubscriptionIndex & index)
+{
+    return LoadIndex(index, 0);
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::LoadIndex(SubscriptionIndex & index, size_t allocateExtraSpace)
 {
     Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
     backingBuffer.Calloc(MaxIndexSize());
+    if (backingBuffer.Get() == nullptr)
+    {
+        mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName());
+        return CHIP_ERROR_NO_MEMORY;
+    }
 
     uint16_t len = static_cast<uint16_t>(MaxIndexSize());
 
     index.mSize = 0;
     if (mStorage->SyncGetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName(), backingBuffer.Get(), len) !=
         CHIP_NO_ERROR)
     {
+        mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName());
         return CHIP_NO_ERROR;
     }
 
     TLV::ScopedBufferTLVReader reader(std::move(backingBuffer), len);
 
-    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    DeleteIndexAndReturnLogErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
     TLV::TLVType arrayType;
-    ReturnErrorOnFailure(reader.EnterContainer(arrayType));
+    DeleteIndexAndReturnLogErrorOnFailure(reader.EnterContainer(arrayType));
 
-    size_t count = 0;
-    CHIP_ERROR err;
-    while ((err = reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag())) == CHIP_NO_ERROR)
+    size_t nodeCount;
+    DeleteIndexAndReturnLogErrorOnFailure(reader.CountRemainingInContainer(&nodeCount));
+    if (nodeCount >= CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
     {
-        if (count >= CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
-        {
-            return CHIP_ERROR_NO_MEMORY;
-        }
+        mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName());
+        return CHIP_ERROR_NO_MEMORY;
+    }
 
+    index.mNodes = std::unique_ptr<ScopedNodeId[]>(new(std::nothrow) ScopedNodeId[nodeCount + allocateExtraSpace]);
+    if (!index.mNodes)
+    {
+        mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionIndex().KeyName());
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    index.mSize = nodeCount;
+
+    for (size_t currentNodeIndex = 0; currentNodeIndex < nodeCount; currentNodeIndex++)
+    {
+        DeleteIndexAndReturnLogErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
         TLV::TLVType containerType;
-        ReturnErrorOnFailure(reader.EnterContainer(containerType));
+        DeleteIndexAndReturnLogErrorOnFailure(reader.EnterContainer(containerType));
```

Comment:
All of these nested `DeleteIndexAndReturnLogErrorOnFailure` could be replaced by making a helper method that does the logic with early returns everywhere, and just have `SimpleSubscriptionResumptionStorage::LoadIndex(SubscriptionIndex & index)` outer call do the index deletion if any error arose in loading the index